lib/compiler/src/beam_ssa_alias.erl

```diff
@@ -0,0 +1,1000 @@
+aa_join_1([], _, State) ->
+    State.
+
+aa_join(#b_var{}=VarA, #b_var{}=VarB, State) ->
+    ARepr = aa_get_representative(VarA, State),
+    BRepr = aa_get_representative(VarB, State),
+    case {ARepr, BRepr} of
+        {Repr, Repr} ->
+            State;
+        _ ->
+            {status, A} = map_get(ARepr, State),
+            {status, B} = map_get(BRepr, State),
+            State#{ ARepr => {status, aa_meet(A, B)}, BRepr => ARepr }
+    end;
+aa_join(_, _, State) ->
+    State.
+
+aa_meet(#b_var{}=Var, SetStatus, State) ->
+    Repr = aa_get_representative(Var, State),
+    {status, Status} = map_get(Repr, State),
+    State#{ Repr => {status, aa_meet(SetStatus, Status)} };
+aa_meet(#b_literal{}, _SetStatus, State) ->
+    State;
+aa_meet([Var|Vars], [Status|Statuses], State) ->
+    aa_meet(Vars, Statuses, aa_meet(Var, Status, State));
+aa_meet([], [], State) ->
+    State.
+
+aa_meet(StatusA, StatusB) ->
+    case {StatusA, StatusB} of
+        {_,aliased} -> aliased;
+        {aliased, _} -> aliased;
+        {unique, unique} -> unique
+    end.
+
+aa_get_status_by_type(Type, StatusByType) ->
+    case StatusByType of
+        #{ Type := Status } ->
+            Status;
+        #{ } ->
+            %% We don't have a direct match, so sequentially try to
+            %% find a matching type.
+            aa_get_status_by_type_ls(Type, aa_sorted_types(StatusByType))
+    end.
+
+aa_get_status_by_type_ls(Type, [{Candidate, Status}|Types]) ->
+    case beam_types:meet(Type, Candidate) of
+        none -> aa_get_status_by_type_ls(Type, Types);
+        _ -> Status
+    end;
+aa_get_status_by_type_ls(_, []) ->
+    aliased.
+
+aa_sorted_types(StatusByType) ->
+    Ls = maps:to_list(StatusByType),
+    lists:sort(fun({#t_float{},_},{#t_number{},_}) ->
+                       true;
+                  ({#t_integer{},_},{#t_number{},_}) ->
+                       true;
+                  ({#t_cons{},_},{#t_list{},_}) ->
+                       true;
+                  ({nil,_},{#t_list{},_}) ->
+                       true;
+                  ({#t_fun{},_},{other,_}) ->
+                       true;
+                  ({#t_map{},_},{other,_}) ->
+                       true;
+                  ({pid,_},{other,_}) ->
+                       true;
+                  ({port,_},{other,_}) ->
+                       true;
+                  ({reference,_},{other,_}) ->
+                       true;
+                  ({#t_bs_matchable{},_},{other,_}) ->
+                       true;
+                  ({#t_bitstring{},_},{#t_bs_matchable{},_}) ->
+                       true;
+                  ({#t_bs_context{},_},{#t_bs_matchable{},_}) ->
+                       true;
+                  ({any,_},{none,_}) ->
+                       true;
+                  ({none,_},{any,_}) ->
+                       false;
+                  ({_,_},{any,_}) ->
+                       true;
+                  ({_,_},{none,_}) ->
+                       true;
```

Comment:
Are we trying to sort the list from "least specific" to "most specific" here? If yes, doesn't this break `if A < B and B < C, then A < C`?

Reply:
The opposite, we want the list to start with the most specific type. And we never want to try a candidate for `any` as long as we have something more specific. So the "end game" should be:

```Erlang
                  %% For handling any and none, we want none at the
                  %% end of the list immediately preceeded by the
                  %% any-typed candidates.
                  ({any,_},{none,_}) ->
                       true;
                  ({none,_},{any,_}) ->
                       false;
                  ({_,_},{_,_}) ->
                       false
```

I see that I also fail to handle cases where the type is `identifier`. will fix that

Reply:
I completely rewrote that part, instead of having a combinatorial mess of different pairwise comparisons I just assigned a rank to each type and sorted on that. Fixup pushed.

Reply:
As @jhogberg  pointed out offline, `aa_get_status_by_type` is not correct. The fix is in 1f35d0df1d717484398a04c9a2c1532f177d0816.